pkg/rpcclient/actor/waiter.go

```diff
@@ -0,0 +1,220 @@
+// its execution result or an error if it's missing from chain after VUB block.
+// It uses optimized web-socket waiter if possible.
+func (a *Actor) waitWithWSWaiter(c RPCEventWaiter, h util.Uint256, vub uint32) (res *state.AppExecResult, waitErr error) {
+	var wsWaitErr error
+	defer func() {
+		if wsWaitErr != nil {
+			res, waitErr = a.waitWithSimpleWaiter(c, h, vub)
+			if waitErr != nil {
+				waitErr = fmt.Errorf("WS waiter error: %w, simple waiter error: %v", wsWaitErr, waitErr)
+			}
+		}
+	}()
+	rcvr := make(chan rpcclient.Notification)
+	defer func() {
+	drainLoop:
+		// Drain rcvr to avoid other notification receivers blocking.
+		for {
+			select {
+			case <-rcvr:
+			default:
+				break drainLoop
+			}
+		}
+		close(rcvr)
+	}()
+	blocksID, err := c.SubscribeForNewBlocksWithChan(nil, rcvr)
+	if err != nil {
+		wsWaitErr = fmt.Errorf("failed to subscribe for new blocks: %w", err)
+		return
+	}
+	defer func() {
+		err = c.Unsubscribe(blocksID)
+		if err != nil {
+			errFmt := "failed to unsubscribe from blocks (id: %s): %v"
+			errArgs := []interface{}{blocksID, err}
+			if waitErr != nil {
+				errFmt += "; wait error: %w"
+				errArgs = append(errArgs, waitErr)
+			}
+			waitErr = fmt.Errorf(errFmt, errArgs...)
+		}
+	}()
+	txsID, err := c.SubscribeForTransactionExecutionsWithChan(nil, rcvr)
+	if err != nil {
+		wsWaitErr = fmt.Errorf("failed to subscribe for execution results: %w", err)
+		return
+	}
+	defer func() {
+		err = c.Unsubscribe(txsID)
+		if err != nil {
+			errFmt := "failed to unsubscribe from transactions (id: %s): %v"
+			errArgs := []interface{}{txsID, err}
+			if waitErr != nil {
+				errFmt += "; wait error: %w"
+				errArgs = append(errArgs, waitErr)
+			}
+			waitErr = fmt.Errorf(errFmt, errArgs...)
+		}
+	}()
+
+	for {
+		select {
+		case ntf := <-rcvr:
+			switch ntf.Type {
+			case neorpc.BlockEventID:
+				block := ntf.Value.(*block.Block)
+				// Execution event follows the block event, thus wait until the block next to the VUB to be sure.
+				if block.Index > vub {
+					waitErr = ErrTxNotAccepted
+					return
+				}
+			case neorpc.ExecutionEventID:
+				aer := ntf.Value.(*state.AppExecResult)
+				if aer.Container.Equals(h) {
+					res = aer
+					return
+				}
+			case neorpc.MissedEventID:
+				// We're toast, retry with non-ws client.
+				wsWaitErr = errors.New("some event was missed")
+				return
+			}
+		case <-c.Context().Done():
+			waitErr = fmt.Errorf("%w: %v", ErrContextDone, c.Context().Err())
+			return
+		}
+	}
+}
+
+// AppExecToInvocation converts state.AppExecResult to result.Invoke and can be used
+// as a wrapper for Wait. The result of AppExecToInvocation doesn't have all fields
+// properly filled, it's limited by State, GasConsumed, Stack, FaultException and Notifications.
+// The result of AppExecToInvocation can be passed to unwrap package helpers.
+func AppExecToInvocation(aer *state.AppExecResult, err error) (*result.Invoke, error) {
```

Comment:
Move it to `result`?

Reply:
Fixed. Althouh, maybe it's better to move it to `unwrap`?

Reply:
It breaks `unwrap.Something` pattern, so `result` is a bit more appropriate, it has `result.Invoke` anyway.